File: ch5/5-4.c
#include <stdio.h>

int strend(char *s, char *t);

int main(void)
{
    char *s1, *s2;
    char s3[] = "first", s4[] = "rst";
    s1 = s3;
    s2 = s4;
    printf("%d\n", strend(s1, s2));
    return 0;
}

int strend(char *s, char *t)
{
    int n = 0;
    for (;*s != '\0'; s++) {
        ;
    }
    for (;*t != '\0'; t++) {
        n++;
    }
    while (n > 0 && (*s-- == *t--)) {
        n--;
    }
    return (n > 0) ? 0 : 1;
}

File: ch5/5-4.go
package main

import "fmt"

func main() {
	s1 := []byte("first")
	s2 := []byte("rst")
	s3 := []byte("foo")
	fmt.Printf("%d\n", strend(s1, s2)) // 1
	fmt.Printf("%d\n", strend(s1, s3)) // 0
}

func strend(s []byte, t []byte) int {
	si, ti := len(s)-1, len(t)-1
	n := ti
	for {
		if n == 0 || s[si] != t[ti] {
			break
		}
		n--
		si--
		ti--
	}
	if n > 0 {
		return 0
	}
	return 1
}
